woocommerce/woocommerce
# Fix/Fix new order triggers when creating/moving orders into draft or from draft to valid statuses

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR resolves issues with new order hook triggers during transitions to and from draft status as well as creating a new order with a draft status, using CPT and HPOS. It fixes the following scenarios:

On CPT

- Placing the order, the new order hook is not triggered ❌
- In the admin, creating a new draft order triggers the new order hook with checkout-draft status ❌
- In the admin, changing an order from draft to completed doesn't trigger the new order hook ❌

On HPOS

- In the admin, creating a new draft order triggers the new order hook with checkout-draft status ❌


Closes #48342 
Closes #48080
Closes #48343 



### How to test the changes in this Pull Request:



### Testing preparation:
1. Install the Debug Log Manager and Code Snippets extensions and add the following snippet:

```php
add_action( 'woocommerce_new_order', 'any', 20, 1 );
function any( $order_id ) {
	$order = wc_get_order( $order_id );
	error_log( 'Order ID: ' . $order_id . ' with status: ' . $order->get_status() );
}
```
2. Keep a tab open with `/wp-admin/tools.php?page=debug-log-manager` and be sure to clear the log before starting the test.

### Testing Steps
1. Clear the error log and go to WooCommerce > Orders and click "Add order"
2. Edit the new order and select "**Draft**" from the status dropdown.
3. Click "Create" and verify that "Created Order ID: xxx with status.. " IS NOT shown on the log manager.
4. Change the Order status again to `Completed` click `Update` and verify that "Created Order ID: xxx with status.. " IS shown on the log manager.
5. Clear the log
6. Make an order using the Checkout Blocks
7. Verify that "Created Order ID: xxx with status.. " IS shown on the log manager.
8. Go to `Admin` > `WooCommerce` > `Advanced` > `Features` and switch between Wordpress posts storage/High-performance order storage (depending on the previous choice). You might have to sync orders before doing so.
<img width="400" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17236129/9b65a3d5-60e7-4769-acba-6a1293d88dfe">

9. Repeat steps 1-8



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
